File: frontend/src/features/vodPipeline/schemas.ts
import { z } from "zod";

/**
 * VOD Pipeline feature Zod schemas.
 *
 * Mirror the actual backend responses from vod_pipeline_api.py. Unknown
 * additional fields are stripped by Zod; missing required fields and
 * unknown enum values surface as controlled ApiErrors instead of React
 * crashes. Status values are permissive strings so a future backend
 * status renders as a neutral badge instead of throwing.
 */

export const KNOWN_VOD_STATUSES = [
  "DISCOVERED",
  "QUEUED",
  "DOWNLOADING",
  "VERIFYING",
  "READY",
  "FAILED",
  "CANCELED",
] as const;

export type KnownVodStatus = (typeof KNOWN_VOD_STATUSES)[number];

/** Permissive status: known values get coloured badges, unknown ones neutral. */
export const vodStatusSchema = z.string();

export const vodProgressSchema = z.object({
  percent: z.number().nullable().optional(),
  downloaded_bytes: z.number().nullable().optional(),
  total_bytes: z.number().nullable().optional(),
  speed_bytes_per_second: z.number().nullable().optional(),
  eta_seconds: z.number().nullable().optional(),
});

export const vodDownloadSchema = z.object({
  started_at: z.string().nullable().optional(),
  completed_at: z.string().nullable().optional(),
  file_name: z.string().nullable().optional(),
  file_size_bytes: z.number().nullable().optional(),
  container: z.string().nullable().optional(),
  duration_seconds: z.number().nullable().optional(),
  width: z.number().nullable().optional(),
  height: z.number().nullable().optional(),
  video_codec: z.string().nullable().optional(),
  audio_codec: z.string().nullable().optional(),
});

export const twitchProfileSchema = z.object({
  schema_version: z.number().optional(),
  id: z.string(),
  login: z.string(),
  channel_url: z.string().optional(),
  display_name: z.string(),
  avatar_url: z.string().nullable().optional(),
  created_at: z.string(),
  updated_at: z.string(),
  last_synced_at: z.string().nullable().optional(),
  // Attached by the API list endpoint; not persisted in the profile file.
  vod_count: z.number().optional(),
});

export const twitchProfileListResponseSchema = z.object({
  profiles: z.array(twitchProfileSchema),
});

export const twitchVodSchema = z.object({
  schema_version: z.number().optional(),
  id: z.string(),
  profile_id: z.string().nullable().optional(),
  twitch_video_id: z.string(),
  library_item_id: z.string().nullable().optional(),
  source_url: z.string(),
  title: z.string(),
  description: z.string(),
  type: z.string(),
  language: z.string(),
  published_at: z.string().nullable().optional(),
  created_at: z.string(),
  duration_seconds: z.number().nullable().optional(),
  thumbnail_url: z.string(),
  view_count: z.number().nullable().optional(),
  status: vodStatusSchema,
  progress: vodProgressSchema,
  download: vodDownloadSchema,
  error: z.string().nullable().optional(),
  updated_at: z.string(),
});

export const vodListResponseSchema = z.object({
  vods: z.array(twitchVodSchema),
});

export const vodSyncResponseSchema = z.object({
  created: z.number(),
  updated: z.number(),
  unchanged: z.number(),
  total: z.number(),
});

export const vodDeleteResponseSchema = z.object({
  id: z.string(),
  deleted: z.boolean(),
});

export const twitchStatusSchema = z.object({
  available: z.boolean(),
  downloader_available: z.boolean().optional(),
  yt_dlp_version: z.string().nullable().optional(),
  ffprobe_available: z.boolean(),
  download_dir_writable: z.boolean(),
  reasons: z.array(z.string()),
  warnings: z.array(z.string()).optional(),
});

export const vodLogResponseSchema = z.object({
  id: z.string(),
  log: z.string(),
});

/** Best-effort error response schema. FastAPI errors return `{detail: ...}`. */
export const apiErrorSchema = z.object({
  detail: z.union([z.string(), z.array(z.any()), z.record(z.string(), z.any())]),
});

File: vod_pipeline/schemas.py
"""Schemas, enums and typed errors for the VOD-pipeline backend core.

This module is the single source of truth for the JSON shapes written to
``twitch_profiles/{profile_id}/profile.json`` and
``vods/{vod_id}/metadata.json``. It deliberately mirrors the conventions
of :mod:`voice_profiles.schemas` but does not import anything from the
voice-profile vertical slice: the VOD-pipeline core must stay isolated.
"""

from __future__ import annotations

from enum import Enum
from typing import Any, Optional

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Schema versioning
# ---------------------------------------------------------------------------

SCHEMA_VERSION = 1
SUPPORTED_SCHEMA_VERSIONS: frozenset[int] = frozenset({1})

# Hard cap for the first sync so we never pull a channel's entire VOD
# history on a single click. Documented and overridable via the service.
DEFAULT_SYNC_LIMIT = 100


# ---------------------------------------------------------------------------
# Enums
# ---------------------------------------------------------------------------

class VodStatus(str, Enum):
    """Lifecycle status of a VOD record."""

    DISCOVERED = "DISCOVERED"
    QUEUED = "QUEUED"
    DOWNLOADING = "DOWNLOADING"
    VERIFYING = "VERIFYING"
    READY = "READY"
    FAILED = "FAILED"
    CANCELED = "CANCELED"


# Transient states: if the worker exits while in one of these, the record
# is automatically marked FAILED by the reaper / startup recovery.
TRANSIENT_VOD_STATUSES = frozenset({
    VodStatus.QUEUED,
    VodStatus.DOWNLOADING,
    VodStatus.VERIFYING,
})

# Statuses from which a download may be (re)started.
STARTABLE_VOD_STATUSES = frozenset({
    VodStatus.DISCOVERED,
    VodStatus.FAILED,
    VodStatus.CANCELED,
})

# Statuses that may be canceled.
CANCELLABLE_VOD_STATUSES = frozenset({
    VodStatus.QUEUED,
    VodStatus.DOWNLOADING,
    VodStatus.VERIFYING,
})

# Terminal statuses (no active worker).
TERMINAL_VOD_STATUSES = frozenset({
    VodStatus.DISCOVERED,
    VodStatus.READY,
    VodStatus.FAILED,
    VodStatus.CANCELED,
})


# Allowed final containers for the downloaded source file. The worker
# picks the best combined format yt-dlp returns; we only ever expose a
# fixed, safe filename to the outside world.
ALLOWED_SOURCE_CONTAINERS = ("mp4", "mkv", "webm", "ts")


# ---------------------------------------------------------------------------
# Typed errors
# ---------------------------------------------------------------------------

class VodError(Exception):
    """Base class for all VOD-pipeline core errors."""


class VodValidationError(VodError):
    """Hard validation failure (bad URL, bad id, bad state transition)."""


class VodNotFoundError(VodError):
    """A VOD with the given id does not exist."""


class VodConflictError(VodError):
    """A VOD-level conflict (already running, wrong state)."""


class VodStorageError(VodError):
    """A persistence-layer failure (corrupt JSON, unknown schema, IO)."""


class TwitchProfileError(Exception):
    """Base class for all Twitch-profile core errors."""


class TwitchProfileValidationError(TwitchProfileError):
    """Hard validation failure (bad login, bad url)."""


class TwitchProfileNotFoundError(TwitchProfileError):
    """A profile with the given id does not exist."""


class TwitchProfileConflictError(TwitchProfileError):
    """A profile-level conflict (duplicate Twitch user id, VODs attached)."""


class TwitchProfileStorageError(TwitchProfileError):
    """A persistence-layer failure (corrupt JSON, unknown schema, IO)."""


class TwitchClientError(Exception):
    """Base class for yt-dlp listing / metadata errors."""


class TwitchNotFoundError(TwitchClientError):
    """A Twitch resource (channel / video / clip) was not found."""


# ---------------------------------------------------------------------------
# Twitch profile / VOD models (used for validation only; persistence is
# plain JSON dicts to stay forward-compatible with additive schema changes)
# ---------------------------------------------------------------------------

class ProfileSchemaVersion(BaseModel):
    schema_version: int = SCHEMA_VERSION


class TwitchProfile(BaseModel):
    """A persisted Twitch profile."""

    schema_version: int = SCHEMA_VERSION
    id: str
    login: str
    channel_url: str = ""
    display_name: str = ""
    created_at: str
    updated_at: str
    last_synced_at: Optional[str] = None


class VodProgress(BaseModel):
    percent: Optional[float] = None
    downloaded_bytes: Optional[int] = None
    total_bytes: Optional[int] = None
    speed_bytes_per_second: Optional[float] = None
    eta_seconds: Optional[float] = None


class VodDownload(BaseModel):
    started_at: Optional[str] = None
    completed_at: Optional[str] = None
    file_name: Optional[str] = None
    file_size_bytes: Optional[int] = None
    container: Optional[str] = None
    duration_seconds: Optional[float] = None
    width: Optional[int] = None
    height: Optional[int] = None
    video_codec: Optional[str] = None
    audio_codec: Optional[str] = None


class VodSchemaVersion(BaseModel):
    schema_version: int = SCHEMA_VERSION


class TwitchVod(BaseModel):
    """A persisted VOD or clip record."""

    schema_version: int = SCHEMA_VERSION
    id: str
    profile_id: str
    twitch_video_id: str  # yt-dlp entry id; stable for dedup
    source_url: str
    title: str = ""
    description: str = ""
    type: str = "archive"  # "archive" (VOD) or "clip"
    language: str = ""
    published_at: Optional[str] = None
    created_at: str
    duration_seconds: Optional[float] = None
    thumbnail_url: str = ""
    view_count: Optional[int] = None
    status: str = VodStatus.DISCOVERED.value
    progress: VodProgress = Field(default_factory=VodProgress)
    download: VodDownload = Field(default_factory=VodDownload)
    error: Optional[str] = None
    updated_at: str


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def empty_progress() -> dict[str, Any]:
    return {
        "percent": None,
        "downloaded_bytes": None,
        "total_bytes": None,
        "speed_bytes_per_second": None,
        "eta_seconds": None,
    }


def empty_download() -> dict[str, Any]:
    return {
        "started_at": None,
        "completed_at": None,
        "file_name": None,
        "file_size_bytes": None,
        "container": None,
        "duration_seconds": None,
        "width": None,
        "height": None,
        "video_codec": None,
        "audio_codec": None,
    }
